File: 380.insert-delete-getrandom-o1/solution.java
class RandomizedSet {
    private ArrayList<Integer> values = new ArrayList<>();
    private Map<Integer, Integer> keys = new HashMap<>();
    
    public boolean insert(int val) {
        if (this.keys.containsKey(val)) {
            return false;
        }
        int n = this.values.size();
        this.keys.put(val, n);
        this.values.add(val);
        return true;
    }
    
    public boolean remove(int val) {
        if (!this.keys.containsKey(val)) {
            return false;
        }
        int idx = this.keys.get(val);
        int lastVal = this.values.get(this.values.size() - 1);
        
        this.values.set(idx, lastVal);
        this.keys.put(lastVal, idx);
        
        this.values.remove(this.values.size() - 1);
        this.keys.remove(val);
        return true;
    }
    
    public int getRandom() {
        Random rand = new Random();
        return this.values.get(rand.nextInt(this.values.size()));
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */

File: 380.insert-delete-getrandom-o1/solution.go
type RandomizedSet struct {
	keys map[int]int
	values []int
	n int
}


func Constructor() RandomizedSet {
	return RandomizedSet{
			keys: make(map[int]int),
			values: make([]int, 0),
			n: 0,
	}
}


func (this *RandomizedSet) Insert(val int) bool {
	if _, containsVal := this.keys[val]; containsVal {
			return false
	}
	
	this.keys[val] = this.n
	this.values = append(this.values, val)
	this.n += 1
	return true
}


func (this *RandomizedSet) Remove(val int) bool {
	if _, containsVal := this.keys[val]; !containsVal {
			return false
	}
	
	lastVal := this.values[this.n - 1]
	valIdx := this.keys[val]
	this.keys[lastVal] = valIdx
	this.values[valIdx] = lastVal
	
	delete(this.keys, val)
	this.values = this.values[:this.n - 1]
	this.n -= 1
	return true
}


func (this *RandomizedSet) GetRandom() int {
	idx := rand.Intn(this.n)
	return this.values[idx]
}


/**
* Your RandomizedSet object will be instantiated and called as such:
* obj := Constructor();
* param_1 := obj.Insert(val);
* param_2 := obj.Remove(val);
* param_3 := obj.GetRandom();
*/